### PR TITLE
Analyzers acquisition experience

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
### Context
This PR implements acquisition flow for custom analyzers.

The process can be described the next way:
A Custom Analyzer library contains a file `Company.AnalyzerTemplate.props`.
image

If this analyzer is referenced from any solution, MSBuild invokes  intrinsic function `RegisterAnalyzer`  during the evaluation. It emits `BuildCheckAcquisitionEventArgs` if the parsed path to custom analyzer is valid.
This event is handled by `BuildCheckConnectorLogger`: it passes the info to `IBuildCheckManager.ProcessAnalyzerAcquisition`.
image

### Tests
Manual and unit tests.
I plan to cover e2e scenarios once MSBuild.Build package with BuildCheck infrastructure is available for public reference (it is needed for custom templates).

### CODE REVIEW

File: src/Build/Evaluation/Expander.cs [hunk 1]
```diff
@@ -3877,6 +3909,15 @@ private bool TryExecuteWellKnownFunction(out object returnVal, object objectInst
                     }
                     else if (_receiverType == typeof(IntrinsicFunctions))
                     {
+                        if (string.Equals(_methodMethodName, nameof(IntrinsicFunctions.RegisterAnalyzer), StringComparison.OrdinalIgnoreCase))
+                        {
+                            if (TryGetArg(args, out string arg0) && _loggingContext != null)
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:3914]: nit - the second check seems much cheaper - so it should be frontloaded

```suggestion
                            if (_loggingContext != null && TryGetArg(args, out string arg0))
```

File: src/Build/Resources/Strings.resx
Comment (JanKrivanek) [inline src/Build/Resources/Strings.resx:1]: Need to be edited manually

**Review by JanKrivanek — COMMENTED** [review]
Will need to do deeper pass for approve/request changes result - but overall looks good.

It feels this deserves a section in documentation

File: src/Build/Evaluation/Expander.cs [hunk 2]
```diff
@@ -311,6 +311,8 @@ private void FlushFirstValueIfNeeded()
 
         private readonly IFileSystem _fileSystem;
 
+        private readonly LoggingContext _loggingContext;
```
[the remark this answers is not in the document]
Comment (YuliiaKovalova) [reply src/Build/Evaluation/Expander.cs:314]: I am not sure if it's a good idea to mention IBuildCheckManager in Expander.
@rainersigwald , what do you think?

File: src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs [hunk 3]
```diff
@@ -2,21 +2,56 @@
...
-        // Acquisition module - https://github.com/dotnet/msbuild/issues/9633
-        return Construct<SharedOutputPathAnalyzer>;
+#if FEATURE_ASSEMBLYLOADCONTEXT
+        /// <summary>
+        /// AssemblyContextLoader used to load DLLs outside of msbuild.exe directory
+        /// </summary>
+        private static readonly CoreClrAssemblyLoader s_coreClrAssemblyLoader = new();
+#endif
+
+        public BuildAnalyzerFactory? CreateBuildAnalyzerFactory(AnalyzerAcquisitionData analyzerAcquisitionData)
```
[the remark this answers is not in the document]
Comment (YuliiaKovalova) [reply src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs:22]: need to discuss it offline

File: src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs [hunk 4]
```diff
@@ -4,19 +4,74 @@
...
+                Assembly? assembly = null;
+#if FEATURE_ASSEMBLYLOADCONTEXT
+                assembly = s_coreClrAssemblyLoader.LoadFromPath(analyzerAcquisitionData.AssemblyPath);
+#else
+                assembly = Assembly.LoadFrom(analyzerAcquisitionData.AssemblyPath);
+#endif
+
+                IEnumerable<Type> analyzerTypes = assembly.GetTypes().Where(t => typeof(BuildAnalyzer).IsAssignableFrom(t));
+
+                if (analyzerTypes.Any())
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs:45]: This (and the foreach below) will leave to double enumearation and call of `GetTypes`. Let's just replace it with the check of emptiness of the `analyzersFactory` list below. Creation of an empty list has negligable cost compared to the acquisition form single package

File: src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs [hunk 5]
```diff
@@ -4,19 +4,74 @@
...
+                {
+                    var analyzersFactory = new List<BuildAnalyzerFactory>();
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs:47]: NIT: should be plural

```suggestion
                    var analyzersFactories = new List<BuildAnalyzerFactory>();
```

File: src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs (hunk 4 above)
Comment (JanKrivanek) [reply src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs:45]: Oh - you are returning empty enumerable below - so then you don't even need to check anything here

File: src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs [hunk 6]
```diff
@@ -24,70 +37,67 @@ public void Initialize(IEventSource eventSource)
...
         {
-            if (buildCheckBuildEventArgs is BuildCheckTracingEventArgs tracingEventArgs)
-            {
-                _stats.Merge(tracingEventArgs.TracingData, (span1, span2) => span1 + span2);
-            }
-            else if (buildCheckBuildEventArgs is BuildCheckAcquisitionEventArgs acquisitionEventArgs)
-            {
-                buildCheckManager.ProcessAnalyzerAcquisition(acquisitionEventArgs.ToAnalyzerAcquisitionData());
-            }
+            handler(e);
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs:70]: I love this - thanks!

File: src/Build/BuildCheck/Infrastructure/IBuildCheckManager.cs [hunk 7]
```diff
@@ -3,52 +3,63 @@
...
 using Microsoft.Build.BackEnd.Logging;
 using Microsoft.Build.BuildCheck.Acquisition;
-using Microsoft.Build.BuildCheck.Infrastructure;
 using Microsoft.Build.BuildCheck.Logging;
 using Microsoft.Build.Framework;
 
-namespace Microsoft.Build.Experimental.BuildCheck;
-
-internal enum BuildCheckDataSource
+namespace Microsoft.Build.Experimental.BuildCheck
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/IBuildCheckManager.cs:11]: nit: files scoped namespace removed here

File: src/Build/Evaluation/Expander.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Evaluation/Expander.cs:314]: My request is controversial (at the very least) and your solution satisfies current requirements - so let's defer the discussion until it's relevant.
Sorry for confusion!

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks good!

I added couple comments - the double iteration in acquisition module is reason for requesting changes, but otherwise it looks good!

**Pull request closed without merging**